### DISCUSSION TITLE
Salvo and sessions?

### DISCUSSION THREAD

**Discussion by jhjacobs81** [salvo-rs/salvo#129] [Q&A]
Hello!

I am trying to build a webapp where one needs to login. so i want to use sessions, and i wonder if anyone had any luck with that?
i came across: https://docs.rs/async-session/latest/async_session/

i can get the example to work, i can initialize a session, put the userid in it. when i print the raw session data i get:
```
Session { id: "a5vjk0tP3qM4CeFJYVWKB8JjCHiLwwahO24yekkB9k8=", expiry: None, data: RwLock { data: {"user_id": "1"}, poisoned: false, .. }, cookie_value: Some("itJ9g6CJ8a2ssRRvO5+b2IHNGt3XJMjA3lUAQhN92Y0eQIoKt7UVBzJ+oIkd5L04lo2eyjor33PW/A9Lctxw9g=="), data_changed: true, destroy: false }
```
So i can safely assume this part of the login process works :-)  But i have no clue about how i would get this to the next page?  This could very well be due to me beeing very new at rust, so hopefully someone more experienced can have a look and point me in the right direction?

**Comment by fergus-hou** [comment]
Try code like this:

```rust
use salvo::extra::session::SessionDepotExt;

#[handler]
async fn my_page(depot: &mut Deport) {
     let session = depot.session();
}
```

**Reply by jhjacobs81** [reply]
i see thats based on async_session as well, so i wasn't far off :)

I'm gonna see if i can work with your suggestion! right now it leads to a whole bunch of other errors, so i need to re-read the documentation i think :) 

i keep running into ``` ^^^^^ could not find `extra` in `salvo` ```

never mind, i fixed that by adding extra to the features in cargo.toml :) im learnign something new every day!

**Comment by jhjacobs81** [comment]
so i have this:
```
#[handler]
pub async fn page_home(res: &mut Response, depot: &mut Depot) {
    let session = depot.session();
    session.insert("user_id", 1);

    let mut pagecontent = String::from("");
    let s = format!("{:?}", session);
    pagecontent.push_str(&s);
    // pagecontent.push_str("orld!<br><br>");
    // pagecontent.push_str(&s);

    // let hello = HomeTemplate { 
        // slug: "hello-world", 
        // title: "Hello world!", 
        // content: &pagecontent
    // };
    res.render(pagecontent);
}
```

but this returns an error, because session.insert(..) expects 1 argument and i supplied 2. that makes no sense.. what use is a session if i cant add data to it? if i uncomment the session.insert line, it prints "None" which i assume is correct since i havent created a new session, so there is nothing to show? i also assume thats why session.insert doesnt work? but how would i go about that? how do i create a session in the depot so that it sticks across requests?

**Reply by chrislearn** [reply]
https://docs.rs/salvo_extra/latest/salvo_extra/session/trait.SessionDepotExt.html
So, your session's type is Option<&Session>. You can not use it as mutable.

Try this:
```rust
let session = depot.session_mut().unwrap();
session.insert("user_id", 1);
```

**Reply by jhjacobs81** [reply]
that gives me:
``` thread 'tokio-runtime-worker' panicked at 'called `Option::unwrap()` on a `None` value' ```
which i assume is because there is no session yet. coming back to my original question about how i would start a new session into the depot :) 

i now have this:
```
#[handler]
pub async fn page_home(res: &mut Response, depot: &mut Depot) {
    let mut session = Session::new();
    depot.set_session(session);
    let new = depot.session_mut().unwrap();
    new.insert("user_id", 1).unwrap();
    let mut pagecontent = String::from("");
    let s = format!("{:?}", new);
    pagecontent.push_str(&s);
    res.render(pagecontent);
}
```

which doesnt give me an error.. so i assume it works. what i expect this to do is:
1. i create a new session
2.  the session gets added to the depot.
3. i add a user id to it

then i create a new handler like so:
```
#[handler]
pub async fn page_news(res: &mut Response, depot: &mut Depot) {
    let new = depot.session_mut().unwrap();
    let mut pagecontent = String::from("");
    let s = format!("{:?}", new);
    pagecontent.push_str(&s);
    res.render(pagecontent);
}
```

which i had assumed will load the original session. but instead it panics. So, i am still not getting it :(

i changed it to this:
```
#[handler]
pub async fn page_news(res: &mut Response, depot: &mut Depot) {
    let new = depot.take_session();
    let mut pagecontent = String::from("");
    let s = format!("{:?}", new);
    pagecontent.push_str(&s);
    res.render(pagecontent);
}
```

and now it doesnt panic anymore, it simply returns None. which leaves me very frustrated :)

**Reply by chrislearn** [reply]
I have created an example: https://github.com/salvo-rs/salvo/blob/main/examples/session-login/src/main.rs

**Reply by jhjacobs81** [reply]
you sir @chrislearn are awesome! :) I got sessions working with your example.
